### ISSUE TITLE
Machine learning model for bot detection

### ISSUE THREAD

**Issue by andreaschandra** [jakartaresearch/adi-buzzer#8]
a given topic or hashtag, we want to see if the population of tweets more likely to flood by buzzer or user organic

or

given a buzzer account, we want to see the major topics to buzzing about

This task includes
- feature engineering (need to do text cleansing, preprocessing)
- baseline model
- early fine-tuning
- evaluation

- [x] define feature set

**Comment by rubentea16** [comment]
Prepare Social Politics Word Dictionary (SPWD)

Propose Feature Set :

- username                   
- name
- is_name_social_political
- desc
- tweets
- n_tweet
- quoted_tweets
- hashtag 
- n_tweet_use_hashtag
- ratio_tweets_use_hashtag
- n_photo
- n_video
- content_url


Feature Engineering :
- [x] is_name_social_political (1/0)  <- create model
- [x] n_tweet
- [x] hashtag
- [x] n_tweet_use_hashtag
- [x] ratio_tweet_use_hashtag
- [x] n_photo
- [x] n_video
- [x] content_url

**Comment by andreaschandra** [comment]
@rubentea16 kalo beragam teknik tapi scorenya masih jelek, mungkin labelingnya kurang konsisten atau kurang banyak

**Comment by rubentea16** [comment]
Current features for create username_contain_politic classifier model

![image](https://user-images.githubusercontent.com/46805454/100617821-9b8adf00-334d-11eb-93a6-efe15f978c4e.png)

**Comment by rubentea16** [comment]
Overview Top 1000 Hashtag buzzer

![image](https://user-images.githubusercontent.com/46805454/100618557-937f6f00-334e-11eb-85ed-528028154b00.png)

**Comment by andreaschandra** [comment]
Baseline model result @rubentea16 

```
BernouliNB
accuracy: 0.78 | precision: 0.60 | recall: 0.21 | f score: 0.32

Linear SVM
accuracy: 0.85 | precision: 0.74 | recall: 0.57 | f score: 0.64

Random Forest
accuracy: 0.82 | precision: 0.74 | recall: 0.43 | f score: 0.54

Gradient Boosting
accuracy: 0.84 | precision: 0.73 | recall: 0.55 | f score: 0.63

AdaBoost
accuracy: 0.81 | precision: 0.63 | recall: 0.58 | f score: 0.60
```

**Comment by rubentea16** [comment]
> Baseline model result @rubentea16
> 
> ```
> BernouliNB
> accuracy: 0.78 | precision: 0.60 | recall: 0.21 | f score: 0.32
> 
> Linear SVM
> accuracy: 0.85 | precision: 0.74 | recall: 0.57 | f score: 0.64
> 
> Random Forest
> accuracy: 0.82 | precision: 0.74 | recall: 0.43 | f score: 0.54
> 
> Gradient Boosting
> accuracy: 0.84 | precision: 0.73 | recall: 0.55 | f score: 0.63
> 
> AdaBoost
> accuracy: 0.81 | precision: 0.63 | recall: 0.58 | f score: 0.60
> ```

ini pake feature apa aja?

**Comment by andreaschandra** [comment]
@rubentea16 tweets aja, cek ini https://github.com/jakartaresearch/adi-buzzer/blob/dev/notebook/40_buzzer_classifier.ipynb

**Comment by rubentea16** [comment]
Benchmark performance use different feature set

https://github.com/jakartaresearch/adi-buzzer/blob/dev/notebook/R3_buzzer_modeling.ipynb

1. Model = RFC(n_estimators=400)
    Features = tweets, user_desc, is_name_social_political, ratio_tweets_use_hashtag, n_tweet, n_photo, n_video
    Word embedding = TFIDF
    Metrics = accuracy: 0.80 | precision: 0.82 | recall: 0.25 | f score: 0.38

2. Model = RFC(n_estimators=400)
    Features = tweets
    Word embedding = TFIDF 
    Metrics = accuracy: 0.80 | precision: 0.81 | recall: 0.24 | f score: 0.37

**Comment by rubentea16** [comment]
## Oversampling training data (Minor class), because of imbalanced dataset

**Oversampling technique: SMOTE (k_neighbors=10)**
Model = RFC(n_estimators=400)
Features = tweets, user_desc, is_name_social_political, ratio_tweets_use_hashtag, n_tweet, n_photo, n_video
Word embedding = TF-IDF
Metrics = accuracy: 0.82 | precision: 0.60 | recall: 0.55 | f score: 0.58

![image](https://user-images.githubusercontent.com/46805454/104816158-ca044100-584b-11eb-8be6-8e34ec7158a0.png)

**Comment by rubentea16** [comment]
## Performance Benchmark

Notes :
- **multiple_feat** = tweets, user_desc, is_name_social_political, ratio_tweets_use_hashtag, n_tweet, n_photo, n_video
- **single_feat** = tweets
- **RFC** = Random Forest Classifier(n_estimator=400)

Model | Desc|Features|Word Embedding|Accuracy|Precision|Recall|F1-score
---|---|---|---|---|---|---|---
RFC|-|multiple-feat|TF-IDF|0.84|0.75|0.33|0.45|
RFC|-|single-feat|TF-IDF|0.84|0.72|0.35|0.47|
SMOTE+RFC|Oversampling train data (Minor class)|multiple-feat|TF-IDF (desc = 3K dim & tweet = 50K dim)|0.86|0.66|0.62|0.64|
SMOTE+RFC|Oversampling train data (Minor class)|single-feat|BPE (tweet = 300 dim)|0.86|0.68|0.57|0.62|
SMOTE+SVC(default)|Oversampling train data (Minor class)|single-feat|BPE (tweet = 300 dim)|0.84|0.59|0.73|0.65|
SMOTE+XGBoost(default)|Oversampling train data (Minor class)|single-feat|BPE (tweet = 300 dim)|0.86|0.66|0.62|0.64|

**Comment by andreaschandra** [comment]
> 0.64

interesting

**Comment by andreaschandra** [comment]
Result after QA label

Algo|acc|pre|rec|fsc
---|---|---|---|---
Bernouli NB | accuracy: 0.78 | precision: 0.75 | recall: 0.21 | f score: 0.33
SVM | accuracy: 0.85 | precision: 0.75 | recall: 0.60 | f score: 0.67
Random Forest | accuracy: 0.81 | precision: 0.77 | recall: 0.34 | f score: 0.47
Gradient Boosting | accuracy: 0.84 | precision: 0.78 | recall: 0.53 | f score: 0.63
AdaBoost | accuracy: 0.82 | precision: 0.67 | recall: 0.56 | f score: 0.61

**Comment by andreaschandra** [comment]
Algo | acc | pre | rec | fsc
---|---|---|---|---
Bernouli NB|accuracy: 0.82 | precision: 0.54 | recall: 0.69 | f score: 0.61
SVM|accuracy: 0.87 | precision: 0.69 | recall: 0.65 | f score: 0.67
RF|accuracy: 0.85 | precision: 0.74 | recall: 0.43 | f score: 0.54
Gradient Boosting | accuracy: 0.87 | precision: 0.72 | recall: 0.54 | f score: 0.62
AdaBoost | accuracy: 0.84 | precision: 0.60 | recall: 0.56 | f score: 0.58
